citest/gcp_testing/gcs_contract.py

```diff
@@ -0,0 +1,193 @@
+
+    Args:
+      bucket: [string] The name of the bucket to list.
+      path: [string] The path within the bucket.
+
+    Returns:
+      An observer that will retrieve the specified content.
+    """
+    return GoogleCloudStorageObserver(
+        self.__gsutil.retrieve, args={'bucket':bucket, 'path':path})
+
+
+class GoogleCloudStorageClauseBuilder(jc.ContractClauseBuilder):
+  """A ContractClause that facilitates observing GCE state."""
+
+  def __init__(self, title, gsutil, retryable_for_secs=0, strict=False):
+    """Construct new clause.
+
+    Args:
+      title: The string title for the clause is only for reporting purposes.
+      gsutil: The GsutilAgent to make the observation for the clause to verify.
+      retryable_for_secs: Number of seconds that observations can be retried
+         if their verification initially fails.
+    """
+    super(GoogleCloudStorageClauseBuilder, self).__init__(
+        title=title, retryable_for_secs=retryable_for_secs)
+    self.__factory = GoogleCloudStorageObjectFactory(gsutil)
+    self.__strict = strict
+
+  def list(self, bucket, path, with_versions=False):
+    """List the bucket contents at a given path.
+
+    Args:
+      bucket: [string] The name of the bucket to list.
+      path: [string] The path within the bucket.
+      with_versions: [boolean] If False then just latest version,
+          otherwise return all versions of all files.
+    """
+    self.observer = self.__factory.new_list(bucket, path, with_versions)
+    title = ('List ' + ('Versioned ' if with_versions else '')
+             + os.path.join(bucket, path))
+    observation_builder = jc.ValueObservationVerifierBuilder(
+        title, strict=self.__strict)
+    self.verifier_builder.append_verifier_builder(observation_builder)
+
+    return observation_builder
+
+  def retrieve(self, bucket, path, no_resource_ok=False):
+    """Retrieve actual contents of file at path.
+
+    Args:
+      bucket: [string] The gcs bucket containing the path.
+      path: [string] The path within the gcs bucket to retreive.
+      no_resource_ok: [bool] True if a 404/resource not found is ok.
+    """
+    self.observer = self.__factory.new_retrieve(bucket, path)
+
+    if no_resource_ok:
+      error_verifier = cli_agent.CliAgentObservationFailureVerifier(
+          title='404 Permitted',
+          error_regex='.*(?:does not exist)|(?:matched no objects).*')
+      disjunction_builder = jc.ObservationVerifierBuilder(
+          'Retrieve {0} bucket={1} path={2} or 404'.format(type, bucket, path))
+      disjunction_builder.append_verifier(error_verifier)
+
+      retrieve_builder = jc.ValueObservationVerifierBuilder(
+          'Retrieve {0} bucket={1} path={2}'.format(type, bucket, path),
+          strict=self.__strict)
+      disjunction_builder.append_verifier_builder(
+          retrieve_builder, new_term=True)
+      self.verifier_builder.append_verifier_builder(
```

Comment:
Added a comment. It would be nice to have internal API support for this. The data model is straight forward, but to get the simple caller API you have to pull this slight of hand.